src/Generator/SchemaMapperGenerator.php

```diff
@@ -374,25 +388,59 @@ protected function generateMapStatementsForObject(Field $root, Variable $payload
                     $optionalVar = $optionalResponseItems[$i];
                 }
 
-                if ($field->isNullable()) {
-                    $ifCondition = $this->builder->funcCall(
-                        'array_key_exists',
-                        [
-                            $field->getName(),
-                            $payloadVariable,
-                        ]
+                if (!$root->hasOneOf()) {
+                    $ifCondition = $field->isNullable()
+                        ? $this->builder->funcCall('array_key_exists', [$field->getName(), $payloadVariable])
+                        : $this->builder->funcCall('isset', [$optionalResponseItems[$i]]);
+
+                    $ifStmt = $this->builder->expr(
+                        $this->builder->methodCall(
+                            $schemaVar,
+                            $this->getSetMethodName($field),
+                            [$optionalVar]
+                        )
                     );
-                } else {
-                    $ifCondition = $this->builder->funcCall('isset', [$optionalResponseItems[$i]]);
+
+                    $statements[] = $this->builder->if($ifCondition, [$ifStmt]);
                 }
-                $ifStmt = $this->builder->expr(
+            }
+
+            if ($root->getDiscriminator()) {
```

Comment:
can you please add in tests example when there's oneOf but no discriminator? because discriminator is optional so it should work even without it...